File: backend/models/action.go
package models

type Action struct {
	ID        string                 `json:"id"`
	DeviceID  string                 `json:"device_id"`
	Type      string                 `json:"type"` // tap, swipe, input, key, open_app
	Params    map[string]interface{} `json:"params"`
	Timestamp int64                  `json:"timestamp"`
	Status    string                 `json:"status"` // pending, executing, done, failed
	Result    string                 `json:"result,omitempty"`
}

type ActionRequest struct {
	DeviceID  string                 `json:"device_id,omitempty"`
	DeviceIDs []string               `json:"device_ids,omitempty"` // For batch operations
	Action    ActionData             `json:"action"`
}

type ActionData struct {
	Type   string                 `json:"type"`
	Params map[string]interface{} `json:"params"`
}

type Macro struct {
	ID          string       `json:"id"`
	Name        string       `json:"name"`
	Description string       `json:"description"`
	Actions     []ActionData `json:"actions"`
	CreatedAt   int64        `json:"created_at"`
}

File: temp_extract/frontend/src/types/action.ts
export interface Action {
    id: string;
    device_id: string;
    type: 'tap' | 'swipe' | 'input' | 'key' | 'open_app' | 'install_apk' | 'push_file';
    params: Record<string, any>;
    timestamp: number;
    status: 'pending' | 'executing' | 'done' | 'failed';
    result?: string;
}

export interface ActionRequest {
    device_id?: string;
    device_ids?: string[]; // For batch operations
    action: {
        type: Action['type'];
        params: Record<string, any>;
    };
}

export interface Macro {
    id: string;
    name: string;
    description: string;
    actions: Omit<Action, 'id' | 'device_id' | 'timestamp' | 'status'>[];
    created_at: number;
}
